core/lib/zksync_core/src/consistency_checker/mod.rs

```diff
@@ -1,204 +1,318 @@
+    async fn transaction_input_data(&self, tx_hash: H256) -> Result<Option<Vec<u8>>, web3::Error>;
+}
+
+#[async_trait]
+impl L1Client for Web3<Http> {
+    async fn transaction_status(&self, tx_hash: H256) -> Result<Option<U64>, web3::Error> {
+        Ok(self
+            .eth()
+            .transaction_receipt(tx_hash)
+            .await?
+            .and_then(|receipt| receipt.status))
+    }
+
+    async fn transaction_input_data(&self, tx_hash: H256) -> Result<Option<Vec<u8>>, web3::Error> {
+        let transaction = self.eth().transaction(TransactionId::Hash(tx_hash)).await?;
+        Ok(transaction.map(|tx| tx.input.0))
+    }
+}
+
+trait UpdateCheckedBatch: fmt::Debug + Send + Sync {
+    fn update_checked_batch(&mut self, last_checked_batch: L1BatchNumber);
+}
+
+/// Default [`UpdateCheckedBatch`] implementation that reports the batch number as a metric.
+impl UpdateCheckedBatch for () {
+    fn update_checked_batch(&mut self, last_checked_batch: L1BatchNumber) {
+        EN_METRICS.last_correct_batch[&CheckerComponent::ConsistencyChecker]
+            .set(last_checked_batch.0.into());
+    }
+}
+
+#[derive(Debug)]
+enum L1DataMismatchBehavior {
+    Bail,
+    Log,
+}
 
 #[derive(Debug)]
 pub struct ConsistencyChecker {
-    // ABI of the zkSync contract
+    /// ABI of the zkSync contract
     contract: ethabi::Contract,
-    // How many past batches to check when starting
+    /// How many past batches to check when starting
     max_batches_to_recheck: u32,
-    web3: Web3<Http>,
-    db: ConnectionPool,
+    sleep_interval: Duration,
+    l1_client: Box<dyn L1Client>,
+    l1_batch_updater: Box<dyn UpdateCheckedBatch>,
+    l1_data_mismatch_behavior: L1DataMismatchBehavior,
+    pool: ConnectionPool,
 }
 
-const SLEEP_DELAY: Duration = Duration::from_secs(5);
-
 impl ConsistencyChecker {
-    pub fn new(web3_url: &str, max_batches_to_recheck: u32, db: ConnectionPool) -> Self {
+    const DEFAULT_SLEEP_INTERVAL: Duration = Duration::from_secs(5);
+
+    pub fn new(web3_url: &str, max_batches_to_recheck: u32, pool: ConnectionPool) -> Self {
         let web3 = Web3::new(Http::new(web3_url).unwrap());
-        let contract = zksync_contracts::zksync_contract();
         Self {
-            web3,
-            contract,
+            contract: zksync_contracts::zksync_contract(),
             max_batches_to_recheck,
-            db,
+            sleep_interval: Self::DEFAULT_SLEEP_INTERVAL,
+            l1_client: Box::new(web3),
+            l1_batch_updater: Box::new(()),
+            l1_data_mismatch_behavior: L1DataMismatchBehavior::Bail,
+            pool,
         }
     }
 
-    async fn check_commitments(&self, batch_number: L1BatchNumber) -> Result<bool, error::Error> {
-        let mut storage = self.db.access_storage().await.unwrap();
+    /// Configures this checker to log discovered L1 batch mismatches instead of bailing with an error.
+    pub fn log_l1_data_mismatches(&mut self) {
```

Comment:
The name provides an impression that it's additive (e.g. it just changes whether logs are emitted or not), while in fact it changes how the `run` function works.
Maybe it will be more clear the other way around, e.g. `bail_on_l1_data_mismatches`?

Additionally, it looks like the main "user" of the checker - EN - has to call this method in order to get the desired behavior, which also seems odd. I'd expect this component to match EN expectations by default.

Reply:
I'm not sure what behavior is intended to be the default one long-term. [The change](https://github.com/matter-labs/zksync-era/pull/758) in behavior is very recent and AFAIU, it's temporary. @RomanBrodetski, could you please clarify whether this "log-only" operation mode will remain long-term?

Reply:
Makes sense; but I'd probably still prefer a more expressive name. Maybe something `do_not_bail_on_l1_data_mismatches` or something of that sort.